File: Algorithm-Templates/LeetCode-Daily/0002-AddTwoNumber/AddTwoNumber.c
#include <stdio.h>
#include <stdlib.h>

// 定义链表节点
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode dummy;//dummy是一个结构体变量，不是指针
    dummy.next = NULL;//dummy的下一个节点就是结果节点的头结点
    struct ListNode* tail = &dummy;//tail指向dummy
    int carry = 0;//carry表示进位

    while (l1 != NULL || l2 != NULL || carry > 0) {
        int v1 =(l1 != NULL) ? l1->val : 0;//v1表示l1的当前位
        int v2 =(l2 != NULL) ? l2->val : 0;//v2表示l2的当前位

        int sum = v1 + v2 + carry;//sum表示当前位的和
        carry = sum / 10;//carry表示进位

        struct ListNode* newNode = (struct ListNode*)malloc(sizeof(struct ListNode));//分配新节点
        newNode->val = sum % 10;//更新新节点的值为sum的个位数
        newNode->next = NULL;//新节点的下一个节点为空

        tail->next = newNode;//把新节点挂到tail的下一个节点
        tail = newNode;//移动tail到新节点

        if (l1 != NULL) l1 = l1->next;//判断是否还有节点，如果非空则移动l1到l1的下一个节点
        if (l2 != NULL) l2 = l2->next;//判断是否还有节点，如果非空则移动l2到l2的下一个节点
    }

    return dummy.next;//返回结果链表的头节点
}

//根据数组创建链表
struct ListNode* createList(int* arr, int size) {
    if (size == 0) return NULL;//如果size为0，则返回空
    struct ListNode* head = (struct ListNode*)malloc(sizeof(struct ListNode));//分配头节点
    head->val = arr[0];//头节点的值为数组的第一个元素
    head->next = NULL;//头节点的下一个节点为空
    struct ListNode* curr = head;//curr表示当前遍历到的节点，先指向头节点
    for (int i = 1; i < size; i++) {//遍历数组
        curr->next = (struct ListNode*)malloc(sizeof(struct ListNode));//分配新节点，并让curr的下一个节点指向新节点
        curr->next->val = arr[i];//新节点的值为数组的当前元素
        curr->next->next = NULL;//新节点的下一个节点为空
        curr = curr->next;//移动curr到新节点
    }
    return head;//返回头节点
}

//打印链表
void printList(struct ListNode* node) {
    while (node) {//遍历链表
        printf("%d", node->val);//打印当前节点的值
        if (node->next) printf("->");//如果当前节点不是尾节点，则打印->
        node = node->next;//移动node到下一个节点
    }
    printf("\n");//打印换行符
}

int main() {//主函数
    int arr1[] = {2, 4, 3};//数组1
    int arr2[] = {5, 6, 4};//数组2
    struct ListNode* l1 = createList(arr1, 3);//根据数组1创建链表1
    struct ListNode* l2 = createList(arr2, 3);//根据数组2创建链表2
    
    printf("List 1: ");
    printList(l1);//打印链表1
    printf("List 2: ");
    printList(l2);//打印链表2
    struct ListNode* result = addTwoNumbers(l1, l2);
    printf("Result: ");
    printList(result);//打印结果链表
    return 0;//返回0
}

File: Algorithm-Templates/LeetCode-Daily/0002-AddTwoNumber/AddTwoNumber.py
from typing import Optional

# 定义链表节点
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# 定义解决方案
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        curr = dummy
        carry = 0
        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0 # 如果l1不为空，则取l1的值，否则取0
            val2 = l2.val if l2 else 0 # 如果l2不为空，则取l2的值，否则取0

            total = val1 + val2 + carry # 计算当前位的和
            carry = total // 10 # 计算进位

            curr.next = ListNode(total % 10) # 创建新节点，值为当前位的和的个位数，并让curr的下一个节点指向新节点
            curr = curr.next # 移动curr到新节点

            l1 = l1.next if l1 else None # 如果l1不为空，则移动l1到l1的下一个节点，否则取None
            l2 = l2.next if l2 else None # 如果l2不为空，则移动l2到l2的下一个节点，否则取None

        return dummy.next # 返回结果链表的头节点

# 根据数组创建链表
def build_list(nodes):
    dummy = ListNode(0) # 创建头节点，哑节点，不表示真实数位，只用来挂结果链表。
    curr = dummy
    for val in nodes:
        curr.next = ListNode(val) # 创建新节点，并让curr的下一个节点指向新节点
        curr = curr.next # 移动curr到新节点
    return dummy.next # 返回结果链表的头节点

# 打印链表
def print_list(node):
    res = [] # 创建一个空列表，用来存储链表的值
    while node:
        res.append(str(node.val)) # 将当前节点的值转换为字符串（str），并添加到列表中（append）
        node = node.next # 移动node到下一个节点
    print("->".join(res)) # 将列表中的所有元素用"->"连接起来（join），并打印出来

if __name__ == "__main__":
    l1 = build_list([2, 4, 3]) # 根据数组[2, 4, 3]创建链表1
    l2 = build_list([5, 6, 4]) # 根据数组[5, 6, 4]创建链表2
    print("List 1: ", end="")
    print_list(l1) # 打印链表1
    print("List 2: ", end="")
    print_list(l2) # 打印链表2
    result = Solution().addTwoNumbers(l1, l2) # 计算两个链表的和
    print("Result: ", end="")
    print_list(result) # 打印结果链表